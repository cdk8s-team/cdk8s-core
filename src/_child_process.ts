/****************************************************************************************
 * Expose `child_process` via our own object that can be easily patched by jest for tests.
 * Consumers of the `child_process` module should add functions to this object and import it
 * wherever needed.
 */
import { spawnSync } from 'child_process';

export const _child_process = {
  spawnSync: spawnSync,
};
