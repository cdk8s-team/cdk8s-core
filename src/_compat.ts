/**
 * Helper functions to allow compatibility with both constructs v3 and v10
 */

import { Construct, IConstruct, Node } from 'constructs';
//
// Interfaces from constructs v3 that do not exist in constructs v10
//

export interface Dependency {
  source: IConstruct;
  target: IConstruct;
};

export interface ValidationError {
  readonly source: Construct;
  readonly message: string;
}

/**
 * Determine whether v10 of `constructs` is being used as a dependency.
 */
function isConstructsV10(node: Node): boolean {
  return typeof((node as any).uniqueId) !== 'string';
}

export function getDependencies(node: Node): Dependency[] {
  if (!isConstructsV10(node)) {
    return node.dependencies as any;
  }

  return node.findAll().map(n => (n as any).node.dependencies.map((ndep: any) => {
    return {
      source: n,
      target: ndep,
    };
  })).flat();
}

export function getValidationErrors(node: Node): string[] {
  if (isConstructsV10(node)) {
    return node.findAll().map((n) => (n as any).node.validate()).flat();
  }

  const errors = node.validate() as any as ValidationError[];
  return errors.map(e => `[${Node.of(e.source).path}] ${e.message}`);
}
