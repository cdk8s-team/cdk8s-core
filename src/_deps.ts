import { SpawnSyncOptions, spawnSync } from 'child_process';

export const _deps = {
  spawnSync: (command: string, args?: ReadonlyArray<string>, options?: SpawnSyncOptions) => spawnSync(command, args, options),
};
