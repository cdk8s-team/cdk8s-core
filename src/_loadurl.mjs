// this is an executable program that downloads a file from a url (with
// redirects) and prints it's output to STDOUT. it is implemented as a program
// because node.js does not support issuing network requests synchronously and
// in the cdk we need everything to be synchronous (it all happens in the
// ctors). so we utilize `spawnSync` to spawn this program as a child process.
// alternatively we could have use `curl` but this is more portable.

import followRedirects from 'follow-redirects';
import { parse } from 'node:url';
import { lstatSync, createReadStream } from 'node:fs';
import process from 'node:process'

const { http, https } = followRedirects;

const url = process.argv[2];
if (!url) {
  console.error(`Usage: ${process.argv[1]} URL`);
  process.exit(1);
}

try {
  if (lstatSync(url).isFile()) {
    createReadStream(url).pipe(process.stdout);
  }
} catch (err) {
  const purl = parse(url);

  if (!purl.protocol) {
    throw new Error(`unable to determine protocol from url: ${url}`);
  }

  const client = getHttpClient(purl.protocol);
  const get = client.get(url, response => {
    if (response.statusCode !== 200) {
      throw new Error(`${response.statusCode} response from http get: ${response.statusMessage}`);
    }

    response.on('data', chunk => process.stdout.write(chunk));
  });

  get.once('error', err => {
    throw new Error(`http error: ${err.message}`);
  });
}

function getHttpClient(protocol) {
  switch (protocol) {
    case 'http:': return http;
    case 'https:': return https;
    default:
      throw new Error(`unsupported protocol "${protocol}" in url: ${url}`);
  }
}
