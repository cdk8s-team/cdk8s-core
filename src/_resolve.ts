import { ApiObject } from './api-object';
import { Lazy } from './lazy';
import { IValueResolver, ResolutionContext } from './value-resolver';

export function resolve(key: string[], value: any, apiObject: ApiObject, resolver?: IValueResolver): any {

  if (value == null) {
    return value;
  }

  // cdk8s token
  if (value instanceof Lazy) {
    const resolved = value.produce();
    return resolve(key, resolved, apiObject, resolver);
  }

  // implicit token as generated by the cdk8s-cli (e.g IntOrString)
  if (typeof (value.resolve) === 'function') {
    const resolved = value.resolve();
    return resolve(key, resolved, apiObject, resolver);
  }

  if (typeof (value) !== 'object') {

    if (resolver) {
      const context = new ResolutionContext(apiObject, key, value);
      resolver.resolve(context);
      return context.newValue;
    } else {
      return value;
    }
  }

  if (Array.isArray(value)) {
    return value.map((x, i) => resolve([...key, `${i}`], x, apiObject, resolver));
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    result[k] = resolve([...key, k], v, apiObject, resolver);
  }

  return result;

}
