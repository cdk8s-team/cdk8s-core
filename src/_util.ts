export interface SanitizeOptions {
  /**
   * Do not include empty objects (no keys).
   * @default false
   */
  readonly filterEmptyObjects?: boolean;

  /**
   * Do not include arrays with no items.
   * @default false
   */
  readonly filterEmptyArrays?: boolean;

  /**
   * Sort dictionary keys.
   * @default true
   */
  readonly sortKeys?: boolean;
}

export function sanitizeValue(obj: any, options: SanitizeOptions = { }): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {

    if (options.filterEmptyArrays && obj.length === 0) {
      return undefined;
    }

    return obj.map(x => sanitizeValue(x, options));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  const sortKeys = options.sortKeys ?? true;
  const keys = sortKeys ? Object.keys(obj).sort() : Object.keys(obj);
  for (const key of keys) {
    const value = obj[key];
    const newValue = sanitizeValue(value, options);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  if (options.filterEmptyObjects && Object.keys(newObj).length === 0) {
    return undefined;
  }

  return newObj;
}

/**
 * Type of a map mapping strings to some arbitrary type
 */
export type Obj<T> = { [key: string]: T };

/**
  * Return whether the given value is an object
  *
  * Even though arrays and instances of classes technically are objects, we
  * usually want to treat them differently, so we return false in those cases.
  */
export function isObject(x: any): x is Obj<any> {
  return x !== null && typeof x === 'object' && !Array.isArray(x)
       && x.constructor.name === 'Object';
}

/**
 * Recursively merge objects together
 *
 * The leftmost object is mutated and returned. Arrays are not merged
 * but overwritten just like scalars.
 *
 * If an object is merged into a non-object, the non-object is lost.
 */
export function deepMerge(...objects: Array<Obj<any> | undefined>) {
  function mergeOne(target: Obj<any>, source: Obj<any>) {
    for (const key of Object.keys(source)) {
      const value = source[key];

      if (isObject(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (!isObject(target[key])) {
          target[key] = value;
        }
        mergeOne(target[key], value);
      } else if (typeof value !== 'undefined') {
        target[key] = value;
      }
    }
  }

  const others = objects.filter(x => x != null) as Array<Obj<any>>;

  if (others.length === 0) { return {}; }
  const into = others.splice(0, 1)[0];

  others.forEach(other => mergeOne(into, other));
  return into;
}
