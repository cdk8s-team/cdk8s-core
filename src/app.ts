import * as fs from 'fs';
import * as path from 'path';
import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { Chart } from './chart';
import { DependencyGraph } from './dependency';
import { Names } from './names';
import { Yaml } from './yaml';

/** The method to divide YAML output into files */
export enum YamlOutputType {
  /** All resources are output into a single YAML file */
  FILE_PER_APP,
  /** Resources are split into seperate files by chart */
  FILE_PER_CHART,
  /** Each resource is output to its own file */
  FILE_PER_RESOURCE,
  /** Each chart in its own folder and each resource in its own file */
  FOLDER_PER_CHART_FILE_PER_RESOURCE,
}

export interface AppProps {
  /**
   * The directory to output Kubernetes manifests.
   *
   * @default - CDK8S_OUTDIR if defined, otherwise "dist"
   */
  readonly outdir?: string;
  /** How to divide the YAML output into files
   * @default YamlOutputType.FILE_PER_CHART
   */
  readonly yamlOutputType?: YamlOutputType;
}

/**
 * Represents a cdk8s application.
 */
export class App extends Construct {
  /**
   * Synthesize a single chart.
   *
   * Each element returned in the resulting array represents a different ApiObject
   * in the scope of the chart.
   *
   * Note that the returned array order is important. It is determined by the various dependencies between
   * the constructs in the chart, where the first element is the one without dependencies, and so on...
   *
   * @returns An array of JSON objects.
   * @param chart the chart to synthesize.
   * @internal
   */
  public static _synthChart(chart: Chart): any[] {

    const app: App = App.of(chart);

    // we must prepare the entire app before synthesizing the chart
    // because the dependency inference happens on the app level.
    resolveDependencies(app);

    // validate the app since we want to call onValidate of the relevant constructs.
    // note this will also call onValidate on constructs from possibly different charts,
    // but thats ok too since we no longer treat constructs as a self-contained synthesis unit.
    validate(app);

    return chartToKube(chart).map(obj => obj.toJson());
  }

  private static of(c: IConstruct): App {

    const scope = Node.of(c).scope;

    if (!scope) {
      // the app is the only construct without a scope.
      return c as App;
    }

    return App.of(scope);
  }

  /**
   * The output directory into which manifests will be synthesized.
   */
  public readonly outdir: string;

  /** How to divide the YAML output into files
   * @default YamlOutputType.FILE_PER_CHART
   */
  public readonly yamlOutputType: YamlOutputType;

  /**
   * Returns all the charts in this app, sorted topologically.
   */
  public get charts(): Chart[] {
    const isChart = (x: IConstruct): x is Chart => x instanceof Chart;
    return new DependencyGraph(Node.of(this))
      .topology()
      .filter(isChart);
  }

  /**
   * Defines an app
   * @param props configuration options
   */
  constructor(props: AppProps = { }) {
    super(undefined as any, '');
    this.outdir = props.outdir ?? process.env.CDK8S_OUTDIR ?? 'dist';
    this.yamlOutputType = props.yamlOutputType ?? YamlOutputType.FILE_PER_CHART;
  }

  /**
   * Synthesizes all manifests to the output directory
   */
  public synth(): void {

    fs.mkdirSync(this.outdir, { recursive: true });

    // Since we plan on removing the distributed synth mechanism, we no longer call `Node.synthesize`, but rather simply implement
    // the necessary operations. We do however want to preserve the distributed validation.
    validate(this);

    // this is kind of sucky, eventually I would like the DependencyGraph
    // to be able to answer this question.
    const hasDependantCharts = resolveDependencies(this);
    const charts = this.charts;

    switch (this.yamlOutputType) {
      case YamlOutputType.FILE_PER_APP:
        let apiObjectList: ApiObject[] = [];

        for (const chart of charts) {
          apiObjectList.push(...chartToKube(chart));
        }

        if (charts.length > 0) {
          Yaml.save(
            path.join(this.outdir, 'app.k8s.yaml'), // There is no "app name", so we just hardcode the file name
            apiObjectList.map((apiObject) => apiObject.toJson()),
          );
        }
        break;

      case YamlOutputType.FILE_PER_CHART:
        const namer: ChartNamer = hasDependantCharts ? new IndexedChartNamer() : new SimpleChartNamer();

        for (const chart of charts) {
          const chartName = namer.name(chart);
          const objects = chartToKube(chart);

          Yaml.save(path.join(this.outdir, chartName), objects.map(obj => obj.toJson()));
        }
        break;

      case YamlOutputType.FILE_PER_RESOURCE:
        for (const chart of charts) {
          const apiObjects = chartToKube(chart);

          apiObjects.forEach((apiObject) => {
            if (!(apiObject === undefined)) {
              const fileName = `${`${apiObject.kind}.${apiObject.metadata.name}`
                .replace(/[^0-9a-zA-Z-_.]/g, '')}.k8s.yaml`;
              Yaml.save(path.join(this.outdir, fileName), [apiObject.toJson()]);
            }
          });
        }
        break;

      case YamlOutputType.FOLDER_PER_CHART_FILE_PER_RESOURCE:
        const folderNamer: ChartNamer = hasDependantCharts ? new IndexedChartFolderNamer() : new SimpleChartFolderNamer();
        for (const chart of charts) {
          const chartName = folderNamer.name(chart);
          const apiObjects = chartToKube(chart);
          const fullOutDir = path.join(this.outdir, chartName);
          fs.mkdirSync(fullOutDir, { recursive: true });

          apiObjects.forEach((apiObject) => {
            if (!(apiObject === undefined)) {
              const fileName = `${`${apiObject.kind}.${apiObject.metadata.name}`
                .replace(/[^0-9a-zA-Z-_.]/g, '')}.k8s.yaml`;
              Yaml.save(path.join(fullOutDir, fileName), [apiObject.toJson()]);
            }
          });
        }
        break;

      default:
        break;
    }

  }

  /**
   * Synthesizes the app into a YAML string.
   *
   * @returns A string with all YAML objects across all charts in this app.
   */
  public synthYaml(): any {
    validate(this);

    const charts = this.charts;
    const docs: any[] = [];

    for (const chart of charts) {
      const apiObjects = chartToKube(chart);
      docs.push(...apiObjects.map(apiObject => apiObject.toJson()));
    }

    return Yaml.stringify(...docs);
  }
}

function validate(app: App) {
  const errors = [];
  for (const child of app.node.findAll()) {
    const childErrors = child.node.validate();
    for (const error of childErrors) {
      errors.push(`[${child.node.path}] ${error}`);
    }
  }

  if (errors.length > 0) {
    throw new Error(`Validation failed with the following errors:\n  ${errors.join('\n  ')}`);
  }
}

function resolveDependencies(app: App) {

  let hasDependantCharts = false;

  const deps = [];
  for (const child of app.node.findAll()) {
    for (const dep of child.node.dependencies) {
      deps.push({ source: child, target: dep });
    }
  }

  for (const dep of deps) {

    // create explicit api object dependencies from implicit construct dependencies
    const targetApiObjects = Node.of(dep.target).findAll().filter(c => c instanceof ApiObject);
    const sourceApiObjects = Node.of(dep.source).findAll().filter(c => c instanceof ApiObject);

    for (const target of targetApiObjects) {
      for (const source of sourceApiObjects) {
        if (target !== source) {
          Node.of(source).addDependency(target);
        }
      }
    }

    // create an explicit chart dependency from implicit construct dependencies
    const sourceChart = Chart.of(dep.source);
    const targetChart = Chart.of(dep.target);

    if (sourceChart !== targetChart) {
      Node.of(sourceChart).addDependency(targetChart);
      hasDependantCharts = true;
    }

  }

  const charts = new DependencyGraph(Node.of(app)).topology()
    .filter(x => x instanceof Chart);

  for (const parentChart of charts) {
    for (const childChart of Node.of(parentChart).children.filter(x => x instanceof Chart)) {
      // create an explicit chart dependency from nested chart relationships
      Node.of(parentChart).addDependency(childChart);
      hasDependantCharts = true;
    }
  }

  return hasDependantCharts;

}

function chartToKube(chart: Chart) {
  return new DependencyGraph(Node.of(chart)).topology()
    .filter(x => x instanceof ApiObject)
    .filter(x => Chart.of(x) === chart) // include an object only in its closest parent chart
    .map(x => (x as ApiObject));
}

interface ChartNamer {
  name(chart: Chart): string;
}

class SimpleChartNamer implements ChartNamer {
  constructor() {
  }

  public name(chart: Chart) {
    return `${Names.toDnsLabel(chart)}.k8s.yaml`;
  }
}

class IndexedChartNamer extends SimpleChartNamer implements ChartNamer {
  private index: number = 0;
  constructor() {
    super();
  }

  public name(chart: Chart) {
    const name = `${this.index.toString().padStart(4, '0')}-${super.name(chart)}`;
    this.index++;
    return name;
  }
}

class SimpleChartFolderNamer implements ChartNamer {
  constructor() {
  }

  public name(chart: Chart) {
    return Names.toDnsLabel(chart);
  }
}

class IndexedChartFolderNamer extends SimpleChartFolderNamer implements ChartNamer {
  private index: number = 0;
  constructor() {
    super();
  }

  public name(chart: Chart) {
    const name = `${this.index.toString().padStart(4, '0')}-${super.name(chart)}`;
    this.index++;
    return name;
  }
}