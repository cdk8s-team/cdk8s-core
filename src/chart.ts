import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { App } from './app';
import { Names } from './names';

const CHART_SYMBOL = Symbol.for('cdk8s.Chart');

export interface ChartProps {
  /**
   * The default namespace for all objects defined in this chart (directly or
   * indirectly). This namespace will only apply to objects that don't have a
   * `namespace` explicitly defined for them.
   *
   * @default - no namespace is synthesized (usually this implies "default")
   */
  readonly namespace?: string;

  /**
   * Labels to apply to all resources in this chart.
   *
   * @default - no common labels
   */
  readonly labels?: { [name: string]: string };

  /**
   * When set to true, the chart will include a special `ConfigMap` resource
   * that contains construct metadata about all other resources in the chart.
   * For example, it will contain a mapping between a resource name and the path
   * of the construct that created it:
   *
   * ```yaml
   * kind: ConfigMap
   * apiVersion: v1
   * data:
   *   myresource__path: Chart/MyResource
   * ```
   *
   * Can also be turned on by setting the `CDK8S_CONSTRUCT_METADATA` environment variable to 'true'.
   *
   * @default false
   */
  readonly constructMetadata?: boolean;
}

export class Chart extends Construct {
  /**
   * Return whether the given object is a Chart.
   *
   * We do attribute detection since we can't reliably use 'instanceof'.
   */
  public static isChart(x: any): x is Chart {
    return x !== null && typeof(x) === 'object' && CHART_SYMBOL in x;
  }

  /**
   * Finds the chart in which a node is defined.
   * @param c a construct node
   */
  public static of(c: IConstruct): Chart {
    if (Chart.isChart(c)) {
      return c;
    }

    const parent = Node.of(c).scope as Construct;
    if (!parent) {
      throw new Error('cannot find a parent chart (directly or indirectly)');
    }

    return Chart.of(parent);
  }

  /**
   * The default namespace for all objects in this chart.
   */
  public readonly namespace?: string;

  /**
   * Chart-level labels.
   */
  private readonly _labels?: { [name: string]: string };

  constructor(scope: Construct, id: string, props: ChartProps = { }) {
    super(scope, id);
    this.namespace = props.namespace;
    this._labels = props.labels ?? {};

    Object.defineProperty(this, CHART_SYMBOL, { value: true });

    const constructMetadata = props.constructMetadata ?? (process.env.CDK8S_CONSTRUCT_METADATA === 'true' ? true : false);

    if (constructMetadata) {
      new ApiObject(this, 'ConstructMetadata', {
        kind: 'ConfigMap',
        apiVersion: 'v1',
        metadata: {
        // this annotation is used by the cli to identify this
        // special resource.
          annotations: { 'cdk8s.io/construct-metadata': 'true' },
        },
        data: {},
      });
    }
  }

  /**
   * Labels applied to all resources in this chart.
   *
   * This is an immutable copy.
   */
  public get labels(): { [name: string]: string } {
    return { ...this._labels };
  }

  /**
   * Generates a app-unique name for an object given it's construct node path.
   *
   * Different resource types may have different constraints on names
   * (`metadata.name`). The previous version of the name generator was
   * compatible with DNS_SUBDOMAIN but not with DNS_LABEL.
   *
   * For example, `Deployment` names must comply with DNS_SUBDOMAIN while
   * `Service` names must comply with DNS_LABEL.
   *
   * Since there is no formal specification for this, the default name
   * generation scheme for kubernetes objects in cdk8s was changed to DNS_LABEL,
   * since itâ€™s the common denominator for all kubernetes resources
   * (supposedly).
   *
   * You can override this method if you wish to customize object names at the
   * chart level.
   *
   * @param apiObject The API object to generate a name for.
   */
  public generateObjectName(apiObject: ApiObject) {
    return Names.toDnsLabel(apiObject);
  }

  /**
   * Create a dependency between this Chart and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders this chart to a set of Kubernetes JSON resources.
   * @returns array of resource manifests
   */
  public toJson(): any[] {
    return App._synthChart(this);
  }
}

