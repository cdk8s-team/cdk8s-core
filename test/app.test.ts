import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { Node, Construct } from 'constructs';
import { Testing, Chart, App, ApiObject, YamlOutputType } from '../src';

test('empty app emits no files', () => {
  // GIVEN
  const app = Testing.app();

  // WHEN
  app.synth();

  // THEN
  expect(fs.readdirSync(app.outdir)).toHaveLength(0);
});

test('app with two charts', () => {
  // GIVEN
  const app = Testing.app();

  // WHEN
  new Chart(app, 'chart1');
  new Chart(app, 'chart2');
  app.synth();

  // THEN
  expect(fs.readdirSync(app.outdir)).toEqual([
    'chart1.k8s.yaml',
    'chart2.k8s.yaml',
  ]);
});

test('app with charts directly dependant', () => {

  // GIVEN
  const app = Testing.app();

  // WHEN
  const chart1 = new Chart(app, 'chart1');
  const chart2 = new Chart(app, 'chart2');
  const chart3 = new Chart(app, 'chart3');

  Node.of(chart1).addDependency(chart2);
  Node.of(chart2).addDependency(chart3);

  app.synth();

  // THEN
  expect(fs.readdirSync(app.outdir)).toEqual([
    '0000-chart3.k8s.yaml',
    '0001-chart2.k8s.yaml',
    '0002-chart1.k8s.yaml',
  ]);

});

test('app with charts indirectly dependant', () => {

  // GIVEN
  const app = Testing.app();

  // WHEN
  const chart1 = new Chart(app, 'chart1');
  const chart2 = new Chart(app, 'chart2');
  const chart3 = new Chart(app, 'chart3');

  const obj1 = new ApiObject(chart1, 'obj1', { apiVersion: 'v1', kind: 'Kind1' });
  const obj2 = new ApiObject(chart2, 'obj2', { apiVersion: 'v1', kind: 'Kind2' });
  const obj3 = new ApiObject(chart3, 'obj3', { apiVersion: 'v1', kind: 'Kind3' });

  Node.of(obj1).addDependency(obj2);
  Node.of(obj2).addDependency(obj3);

  app.synth();

  // THEN
  expect(fs.readdirSync(app.outdir)).toEqual([
    '0000-chart3.k8s.yaml',
    '0001-chart2.k8s.yaml',
    '0002-chart1.k8s.yaml',
  ]);

});

test('default output directory is "dist"', () => {
  // GIVEN
  const prev = process.cwd();
  const workdir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk8s-'));

  try {
    process.chdir(workdir);

    // WHEN
    const app = new App();
    new Chart(app, 'chart1');
    app.synth();

    // THEN
    expect(app.outdir).toEqual('dist');
    expect(fs.existsSync('./dist')).toBeTruthy();
    expect(fs.statSync('./dist').isDirectory());
    expect(fs.existsSync('./dist/chart1.k8s.yaml')).toBeTruthy();
  } finally {
    process.chdir(prev);
    // fs.rmdirSync(workdir, { recursive: true });
  }
});

test('app with dependent and independent charts', () => {

  // GIVEN
  const app = Testing.app();

  // WHEN
  const chart1 = new Chart(app, 'chart1');
  new Chart(app, 'chart2');
  const chart3 = new Chart(app, 'chart3');
  new Chart(app, 'chart4');

  Node.of(chart1).addDependency(chart3);

  app.synth();

  // THEN
  expect(fs.readdirSync(app.outdir)).toEqual([
    '0000-chart3.k8s.yaml',
    '0001-chart1.k8s.yaml',
    '0002-chart2.k8s.yaml',
    '0003-chart4.k8s.yaml',
  ]);

});

test('app with chart dependencies via custom constructs', () => {

  class CustomConstruct extends Construct {

    public obj: ApiObject;

    constructor(scope: Construct, id: string) {
      super(scope, id);

      this.obj = new ApiObject(this, `${id}obj`, { apiVersion: 'v1', kind: 'CustomConstruct' });
    }
  }

  const app = Testing.app();
  const chart1 = new Chart(app, 'chart1');
  const chart2 = new Chart(app, 'chart2');

  const microService = new CustomConstruct(chart1, 'MicroService');
  const dataBase = new CustomConstruct(chart2, 'DataBase');

  Node.of(microService).addDependency(dataBase);

  app.synth();

  expect(fs.readdirSync(app.outdir)).toEqual([
    '0000-chart2.k8s.yaml',
    '0001-chart1.k8s.yaml',
  ]);

});

test('app with nested charts will deduplicate api objects', () => {

  class CustomConstruct extends Construct {

    public obj: ApiObject;

    constructor(scope: Construct, id: string) {
      super(scope, id);

      this.obj = new ApiObject(this, `${id}obj`, { apiVersion: 'v1', kind: 'CustomConstruct' });
    }
  }

  const app = Testing.app();
  const chart = new Chart(app, 'chart1');
  const childChart = new Chart(chart, 'chart2');
  new CustomConstruct(chart, 'child1');
  new CustomConstruct(childChart, 'child2');

  app.synth();

  expect(fs.readdirSync(app.outdir)).toEqual([
    '0000-chart1-chart2-c883b207.k8s.yaml',
    '0001-chart1.k8s.yaml',
  ]);
  expect(fs.readFileSync(path.join(app.outdir, '0000-chart1-chart2-c883b207.k8s.yaml'), 'utf8')).toMatchSnapshot();
  expect(fs.readFileSync(path.join(app.outdir, '0001-chart1.k8s.yaml'), 'utf8')).toMatchSnapshot();

});

test('app with nested charts will deduplicate api objects (using custom classes)', () => {

  class ChildChart1 extends Chart {
    constructor(scope: Construct, name: string) {
      super(scope, name);
      new ApiObject(this, 'namespace1', { apiVersion: 'v1', kind: 'Namespace' });
    }
  }

  class ChildChart2 extends Chart {
    constructor(scope: Construct, name: string) {
      super(scope, name);
      new ApiObject(this, 'namespace2', { apiVersion: 'v1', kind: 'Namespace' });
    }
  }

  class ParentChart extends Chart {
    constructor(scope: Construct, name: string) {
      super(scope, name);
      new ChildChart1(this, 'child1');
      new ChildChart2(this, 'child2');
      new ApiObject(this, 'namespace3', { apiVersion: 'v1', kind: 'Namespace' });
    }
  }

  const app = Testing.app();
  new ParentChart(app, 'parent');

  app.synth();

  expect(fs.readdirSync(app.outdir)).toEqual([
    '0000-parent-child1-c8e38b2d.k8s.yaml',
    '0001-parent-child2-c882caae.k8s.yaml',
    '0002-parent.k8s.yaml',
  ]);
  expect(fs.readFileSync(path.join(app.outdir, '0000-parent-child1-c8e38b2d.k8s.yaml'), 'utf8')).toMatchSnapshot();
  expect(fs.readFileSync(path.join(app.outdir, '0001-parent-child2-c882caae.k8s.yaml'), 'utf8')).toMatchSnapshot();
  expect(fs.readFileSync(path.join(app.outdir, '0002-parent.k8s.yaml'), 'utf8')).toMatchSnapshot();

});

test('synth calls validate', () => {

  class ValidatingConstruct extends Construct {

    public validateInvoked = false;

    constructor(scope: Construct, id: string) {
      super(scope, id);
      this.node.addValidation({
        validate: (): string[] => { this.validateInvoked = true; return []; },
      });
    }

  }

  const app = new App();

  const construct = new ValidatingConstruct(app, 'ValidatingConstruct');

  app.synth();

  expect(construct.validateInvoked).toBeTruthy();

});

test('apps with varying yamlOutputTypes; two charts, no objects', () => {
  const testSpecs = [
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_CHART },
      result: ['chart1.k8s.yaml', 'chart2.k8s.yaml'],
    },
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_APP },
      result: ['app.k8s.yaml'],
    },
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_RESOURCE },
      result: [],
    },
  ];
  for (const testSpec of testSpecs) {
    // GIVEN
    const app = Testing.app(testSpec.props);

    // WHEN
    new Chart(app, 'chart1');
    new Chart(app, 'chart2');
    app.synth();

    // THEN
    expect(fs.readdirSync(app.outdir)).toEqual(testSpec.result);
  }
});

test('return app as yaml string', () => {
  // GIVEN
  const app = Testing.app();

  // WHEN
  const chart = new Chart(app, 'chart1');
  new ApiObject(chart, 'obj1', { apiVersion: 'v1', kind: 'Kind1' });
  new ApiObject(chart, 'obj2', { apiVersion: 'v1', kind: 'Kind2' });

  new Chart(app, 'chart2');

  const chart3 = new Chart(app, 'chart3');
  new ApiObject(chart3, 'obj3', { apiVersion: 'v1', kind: 'Kind3' });
  new ApiObject(chart3, 'obj4', { apiVersion: 'v1', kind: 'Kind4' });

  const a = app.synthYaml();

  // THEN
  expect(a).toMatchSnapshot();
},
);

test('apps with varying yamlOutputTypes; charts indirectly dependant, multiple objects', () => {
  const testSpecs = [
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_CHART },
      result: [
        '0000-chart3.k8s.yaml',
        '0001-chart2.k8s.yaml',
        '0002-chart1.k8s.yaml',
      ],
    },
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_APP },
      result: ['app.k8s.yaml'],
    },
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_RESOURCE },
      result: [
        'Kind1.chart1-obj1-c818e77f.k8s.yaml',
        'Kind2.chart2-obj2-c8636f20.k8s.yaml',
        'Kind3.chart3-obj3-c8abbfb5.k8s.yaml',
      ],
    },
  ];

  for (const testSpec of testSpecs) {
    // GIVEN
    const app = Testing.app(testSpec.props);

    // WHEN
    const chart1 = new Chart(app, 'chart1');
    const chart2 = new Chart(app, 'chart2');
    const chart3 = new Chart(app, 'chart3');

    const obj1 = new ApiObject(chart1, 'obj1', { apiVersion: 'v1', kind: 'Kind1' });
    const obj2 = new ApiObject(chart2, 'obj2', { apiVersion: 'v1', kind: 'Kind2' });
    const obj3 = new ApiObject(chart3, 'obj3', { apiVersion: 'v1', kind: 'Kind3' });

    Node.of(obj1).addDependency(obj2);
    Node.of(obj2).addDependency(obj3);

    app.synth();

    // THEN
    expect(fs.readdirSync(app.outdir)).toEqual(testSpec.result);
  }
});

test('apps with varying yamlOutputTypes; chart dependencies via custom constructs', () => {

  class CustomConstruct extends Construct {

    public obj: ApiObject;

    constructor(scope: Construct, id: string) {
      super(scope, id);

      this.obj = new ApiObject(this, `${id}obj`, { apiVersion: 'v1', kind: 'CustomConstruct' });
    }
  }

  const testSpecs = [
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_CHART },
      result: ['0000-chart2.k8s.yaml', '0001-chart1.k8s.yaml'],
    },
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_APP },
      result: ['app.k8s.yaml'],
    },
    {
      props: { yamlOutputType: YamlOutputType.FILE_PER_RESOURCE },
      result: [
        'CustomConstruct.chart1-microservice-microserviceobj-c8e1164f.k8s.yaml',
        'CustomConstruct.chart2-database-databaseobj-c8b5eba3.k8s.yaml',
      ],
    },
  ];

  for (const testSpec of testSpecs) {
    const app = Testing.app(testSpec.props);
    const chart1 = new Chart(app, 'chart1');
    const chart2 = new Chart(app, 'chart2');

    const microService = new CustomConstruct(chart1, 'MicroService');
    const dataBase = new CustomConstruct(chart2, 'DataBase');

    Node.of(microService).addDependency(dataBase);

    app.synth();

    expect(fs.readdirSync(app.outdir)).toEqual(testSpec.result);
  }
});