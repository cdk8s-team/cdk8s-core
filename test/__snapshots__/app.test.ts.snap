// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app with nested charts will deduplicate api objects (with inheritance) 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: parent-child1-namespace1-c871643e
"
`;

exports[`app with nested charts will deduplicate api objects (with inheritance) 2`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: parent-child2-namespace2-c806260b
"
`;

exports[`app with nested charts will deduplicate api objects (with inheritance) 3`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: parent-namespace3-c8bf842a
"
`;

exports[`app with nested charts will deduplicate api objects 1`] = `
"apiVersion: v1
kind: CustomConstruct
metadata:
  name: chart1-chart2-child2-child2obj-c828dca6
"
`;

exports[`app with nested charts will deduplicate api objects 2`] = `
"apiVersion: v1
kind: CustomConstruct
metadata:
  name: chart1-child1-child1obj-c868628e
"
`;
